File: src/Java/1. First Course/2. First Hole/src/main/refactoringgolf/store/Order.java
package refactoringgolf.store;

import java.util.Date;
import java.util.HashSet;
import java.util.Set;

public class Order {

	private Customer customer;
	private Salesman salesman;
	private Date orderedOn;
	private String deliveryStreet;
	private String deliveryCity;
	private String deliveryCountry;
	private Set<OrderItem> items;

	public Order(Customer customer, Salesman salesman, String deliveryStreet, String deliveryCity, String deliveryCountry, Date orderedOn) {
		this.customer = customer;
		this.salesman = salesman;
		this.deliveryStreet = deliveryStreet;
		this.deliveryCity = deliveryCity;
		this.deliveryCountry = deliveryCountry;
		this.orderedOn = orderedOn;
		this.items = new HashSet<OrderItem>();
	}

	public Customer getCustomer() {
		return customer;
	}

	public Salesman getSalesman() {
		return salesman;
	}

	public Date getOrderedOn() {
		return orderedOn;
	}

	public String getDeliveryStreet() {
		return deliveryStreet;
	}

	public String getDeliveryCity() {
		return deliveryCity;
	}

	public String getDeliveryCountry() {
		return deliveryCountry;
	}

	public Set<OrderItem> getItems() {
		return items;
	}

	public float total() {
		float totalItems = totalItems();
		float tax = tax(totalItems);
		int shipping = shipping();

		return totalItems + tax + shipping;
	}

	private int shipping() {
		int shipping = 15;
		if (this.deliveryCountry == "USA") {
			shipping = 0;
		}
		return shipping;
	}
	
	private float tax(float totalAmount) {
		return totalAmount * 5 / 100;
	}

	private float totalItems() {
		float totalAmount = 0;
		for (OrderItem item : items) {
			float discount=0;
			float itemAmount = item.getProduct().getUnitPrice() * item.getQuantity();
			if (item.getProduct().getCategory() == ProductCategory.Accessories) {
				discount = 0;
				if (itemAmount >= 100) {
					discount = itemAmount * 10 / 100;
				}
			}
			if (item.getProduct().getCategory() == ProductCategory.Bikes) {
				discount = itemAmount * 20 / 100;
			}
			if (item.getProduct().getCategory() == ProductCategory.Cloathing) {
				discount = 0;
				if (item.getQuantity() > 2) {
					discount = item.getProduct().getUnitPrice();
				}
			}
			itemAmount = itemAmount - discount;
			totalAmount += itemAmount;
		}
		return totalAmount;
	}
}

File: src/Net/1. First Course/5. Fourth Hole/RefactoringGolf.Store/Order.cs
namespace RefactoringGolf.Store
{
    using System;
    using System.Collections.Generic;

    public class Order
    {
        public Customer Customer { get; private set; }

        public Salesman Salesman { get; private set; }

        public DateTime OrderedOn { get; private set; }

        public string DeliveryStreet { get; private set; }

        public string DeliveryCity { get; private set; }

        public string DeliveryCountry { get; private set; }

        public IList<OrderItem> Items { get; private set; }

        public Order(Customer customer, Salesman salesman, string deliveryStreet, string deliveryCity, string deliveryCountry, DateTime orderedOn)
        {
            Customer = customer;
            Salesman = salesman;
            DeliveryStreet = deliveryStreet;
            DeliveryCity = deliveryCity;
            DeliveryCountry = deliveryCountry;
            OrderedOn = orderedOn;
            Items = new List<OrderItem>();
        }

        public decimal Total()
        {
            var totalItems = this.TotalItems();
            var tax = this.Tax(totalItems);
            var shipping = this.Shipping();

            return totalItems + tax + shipping;
        }

        private int Shipping()
        {
            int shipping = 15;
            if (this.DeliveryCountry == "USA")
            {
                shipping = 0;
            }
            return shipping;
        }

        private decimal Tax(decimal totalItems)
        {
            var tax = totalItems * 5 / 100;
            return tax;
        }

        private decimal TotalItems()
        {
            decimal totalItems = 0;
            foreach (var item in this.Items)
            {
                totalItems += item.Total();
            }
            return totalItems;
        }
    }
}